### PR TITLE
wrap: Add patch_directory support

### PR DESCRIPTION
**Pull request by xclaesse**
I'm just wondering if we should force patch_directory inside `packagefiles` subdir?

### CODE REVIEW

**Comment by mensinda** [comment]
> I'm just wondering if we should force patch_directory inside `packagefiles` subdir?

Yes, I would say we should do that. But then I would recommend renaming `packagefiles` to something diffrerent (`localresources` maybe? Naming things is hard).

**Comment by xclaesse** [comment]
I agree the naming is a bit unfortunate, but I don't think it's worth adding yet another directory.

**Comment by xclaesse** [comment]
PR updated to use packagefiles directory.

**Comment by mensinda** [comment]
Rather than adding a new directory, I wanted to suggest to rename the existing `packagefiles` dir, since it hasn't been released yet.

**Comment by xclaesse** [comment]
oh, forgot that it was not yet released, good point.

**Comment by xclaesse** [comment]
@mensinda tbh, I don't think the naming is that bad. One advantage is that `packagecache` and `packagefiles` are sorted together in when listing dirs in subprojects/. Any objection on merging this PR?

**Comment by mensinda** [comment]
Implementation wise, this PR is good to go. I also don't have any strong feelings regarding the directory name (directory sorting is indeed a good point).

Since this is technically a new feature I would say that @jpakkane should have the last word whether `packagefiles` should stay or not.

Also, it would be a good idea to explicitly document that this *only* works for patches. Otherwise, we are likely to get bug reports :)

File: docs/markdown/snippets/wrap_patch.md [hunk 1]
```diff
@@ -4,3 +4,9 @@ It is now possible to use the `patch_filename` and `source_filename` value in a
 `.wrap` file without `*_url` to specify a local source / patch file. All local
 files must be located in the `subprojects/packagefiles` directory. The `*_hash`
 entries are optional with this setup.
+
+## Local wrap patch directory
+
+Wrap files can now specify `patch_directory` instead of `patch_filename` in the
+case overlay files are local. The patch directory must be placed in
+`subprojects/packagefiles` directory.
```
Comment (jpakkane) [inline docs/markdown/snippets/wrap_patch.md:12]: "Overlay" is not an immediately obvious term. This should explain briefly what it does (e.g. "copies files in this dir unaltered over the source dir").

Comment (xclaesse) [reply docs/markdown/snippets/wrap_patch.md]: done.

Comment (xclaesse) [reply docs/markdown/snippets/wrap_patch.md]: Note that the doc already uses the term 'overlay' for this, see https://mesonbuild.com/Wrap-dependency-system-manual.html#specific-to-wrapfile. But I agree it's better to clarify in the release snippet too.

File: mesonbuild/wrap/wrap.py [hunk 2]
```diff
@@ -422,13 +418,22 @@ def get_file_internal(self, what: str) -> str:
...
     def apply_patch(self) -> None:
-        path = self.get_file_internal('patch')
-        try:
-            shutil.unpack_archive(path, self.subdir_root)
-        except Exception:
-            with tempfile.TemporaryDirectory() as workdir:
-                shutil.unpack_archive(path, workdir)
-                self.copy_tree(workdir, self.subdir_root)
+        if 'patch_filename' in self.wrap.values:
+            path = self.get_file_internal('patch')
```
Comment (jpakkane) [inline mesonbuild/wrap/wrap.py:422]: This should check if `patch_directory` is also defined and error out if it is. Otherwise the file is ambiguous.

Comment (xclaesse) [reply mesonbuild/wrap/wrap.py:425]: Good point, fixed now.

**Pull request merged**